File: Server~/src/tools/physicsAndComponentTools.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Raycast Query Tool
const raycastToolName = 'raycast_query';
const raycastToolDescription = 'Executes a physics raycast query in the Unity scene and returns hit information';

const vector3Schema = z.object({
  x: z.number(),
  y: z.number(),
  z: z.number()
});

const raycastParamsSchema = z.object({
  origin: vector3Schema.describe("Origin point of the raycast"),
  direction: vector3Schema.describe("Direction vector of the raycast"),
  maxDistance: z.number().optional().describe("Maximum distance to check (default: 1000)"),
  layerMask: z.number().int().optional().describe("Layer mask bitmask (default: all layers)")
});

// Copy Component Tool
const copyComponentToolName = 'copy_component';
const copyComponentDescription = 'Copies a component from a source GameObject to a target GameObject, preserving its field values';

const copyComponentParamsSchema = z.object({
  componentType: z.string().describe("The name of the component class to copy (e.g. 'BoxCollider')"),
  sourceInstanceId: z.number().int().optional().describe("The instance ID of the source GameObject"),
  sourceObjectPath: z.string().optional().describe("The hierarchy path of the source GameObject (e.g. 'Player')"),
  targetInstanceId: z.number().int().optional().describe("The instance ID of the target GameObject"),
  targetObjectPath: z.string().optional().describe("The hierarchy path of the target GameObject (e.g. 'Enemy')")
});

export function registerPhysicsAndComponentTools(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${raycastToolName}`);
  server.tool(
    raycastToolName,
    raycastToolDescription,
    raycastParamsSchema.shape,
    async (params: any): Promise<CallToolResult> => {
      try {
        logger.info(`Executing tool: ${raycastToolName}`, params);
        const result = await raycastQueryHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${raycastToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${raycastToolName}`, error);
        throw error;
      }
    }
  );

  logger.info(`Registering tool: ${copyComponentToolName}`);
  server.tool(
    copyComponentToolName,
    copyComponentDescription,
    copyComponentParamsSchema.shape,
    async (params: any): Promise<CallToolResult> => {
      try {
        logger.info(`Executing tool: ${copyComponentToolName}`, params);
        const result = await copyComponentHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${copyComponentToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${copyComponentToolName}`, error);
        throw error;
      }
    }
  );
}

async function raycastQueryHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.origin || !params.direction) {
    throw new McpUnityError(ErrorType.VALIDATION, "Parameters 'origin' and 'direction' are required");
  }

  const response = await mcpUnity.sendRequest({
    method: raycastToolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to execute raycast'
    );
  }

  let text = "Raycast did not hit any collider.";
  if (response.hit) {
    text = `Raycast hit GameObject '${response.gameObjectName}' at distance ${response.distance}m. Hit point: (${response.point.x}, ${response.point.y}, ${response.point.z})`;
  }

  return {
    content: [{
      type: 'text' as const,
      text
    }],
    data: response
  };
}

async function copyComponentHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.componentType) {
    throw new McpUnityError(ErrorType.VALIDATION, "Parameter 'componentType' is required");
  }
  if (!params.sourceInstanceId && !params.sourceObjectPath) {
    throw new McpUnityError(ErrorType.VALIDATION, "Either 'sourceInstanceId' or 'sourceObjectPath' must be provided");
  }
  if (!params.targetInstanceId && !params.targetObjectPath) {
    throw new McpUnityError(ErrorType.VALIDATION, "Either 'targetInstanceId' or 'targetObjectPath' must be provided");
  }

  const response = await mcpUnity.sendRequest({
    method: copyComponentToolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to copy component'
    );
  }

  return {
    content: [{
      type: 'text' as const,
      text: response.message || 'Successfully copied component'
    }],
    data: {
      action: response.action
    }
  };
}

File: Editor/Tools/PhysicsAndComponentTools.cs
using System;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for executing a Physics Raycast in the scene
    /// </summary>
    public class RaycastQueryTool : McpToolBase
    {
        public RaycastQueryTool()
        {
            Name = "raycast_query";
            Description = "Executes a physics raycast query in the Unity scene and returns hit information";
        }

        public override JObject Execute(JObject parameters)
        {
            JObject originObj = parameters["origin"] as JObject;
            JObject directionObj = parameters["direction"] as JObject;
            float maxDistance = parameters["maxDistance"]?.ToObject<float>() ?? 1000f;
            int layerMask = parameters["layerMask"]?.ToObject<int>() ?? Physics.AllLayers;

            if (originObj == null || directionObj == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Parameters 'origin' and 'direction' containing 'x', 'y', 'z' must be provided",
                    "validation_error"
                );
            }

            try
            {
                Vector3 origin = new Vector3(
                    originObj["x"]?.ToObject<float>() ?? 0f,
                    originObj["y"]?.ToObject<float>() ?? 0f,
                    originObj["z"]?.ToObject<float>() ?? 0f
                );

                Vector3 direction = new Vector3(
                    directionObj["x"]?.ToObject<float>() ?? 0f,
                    directionObj["y"]?.ToObject<float>() ?? 0f,
                    directionObj["z"]?.ToObject<float>() ?? 0f
                );

                if (direction == Vector3.zero)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "Raycast direction cannot be zero vector",
                        "validation_error"
                    );
                }

                direction.Normalize();

                // Editor transform edits (move_gameobject, set_transform, etc.) don't automatically
                // push updated colliders into the physics world until the next physics step - which
                // in Edit mode may never come. Without this, a raycast run immediately after moving
                // an object can hit its stale pre-move position.
                Physics.SyncTransforms();

                bool hit = Physics.Raycast(origin, direction, out RaycastHit hitInfo, maxDistance, layerMask);

                if (hit)
                {
                    return new JObject
                    {
                        ["success"] = true,
                        ["hit"] = true,
                        ["distance"] = hitInfo.distance,
                        ["point"] = new JObject
                        {
                            ["x"] = hitInfo.point.x,
                            ["y"] = hitInfo.point.y,
                            ["z"] = hitInfo.point.z
                        },
                        ["normal"] = new JObject
                        {
                            ["x"] = hitInfo.normal.x,
                            ["y"] = hitInfo.normal.y,
                            ["z"] = hitInfo.normal.z
                        },
                        ["gameObjectName"] = hitInfo.collider.gameObject.name,
                        ["gameObjectPath"] = GetGameObjectPath(hitInfo.collider.gameObject),
                        // Public ID scheme shared with every other tool - a raw GetInstanceID()
                        // is not resolvable by GameObjectResolver under MCP_UNITY_ENTITY_ID_API.
                        ["instanceId"] = UnityObjectId.GetObjectId(hitInfo.collider.gameObject)
                    };
                }
                else
                {
                    return new JObject
                    {
                        ["success"] = true,
                        ["hit"] = false,
                        ["message"] = "Raycast did not hit any collider."
                    };
                }
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error executing raycast: {ex.Message}",
                    "raycast_error"
                );
            }
        }

        private string GetGameObjectPath(GameObject obj)
        {
            string path = obj.name;
            while (obj.transform.parent != null)
            {
                obj = obj.transform.parent.gameObject;
                path = obj.name + "/" + path;
            }
            return path;
        }
    }

    /// <summary>
    /// Tool to copy component values/setup from a source GameObject to a target GameObject
    /// </summary>
    public class CopyComponentTool : McpToolBase
    {
        public CopyComponentTool()
        {
            Name = "copy_component";
            Description = "Copies a component from a source GameObject to a target GameObject, preserving its field values";
        }

        public override JObject Execute(JObject parameters)
        {
            string componentType = parameters["componentType"]?.ToObject<string>();

            if (string.IsNullOrEmpty(componentType))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Parameter 'componentType' must be provided",
                    "validation_error"
                );
            }

            try
            {
                GameObjectResolver.Result sourceResult = ResolveGameObject(parameters, "Source", "sourceInstanceId", "sourceObjectPath");
                if (sourceResult.Error != null) return sourceResult.Error;

                GameObjectResolver.Result targetResult = ResolveGameObject(parameters, "Target", "targetInstanceId", "targetObjectPath");
                if (targetResult.Error != null) return targetResult.Error;

                GameObject sourceObj = sourceResult.GameObject;
                GameObject targetObj = targetResult.GameObject;

                Component sourceComp = sourceObj.GetComponent(componentType);
                if (sourceComp == null)
                {
                    // Try finding component with case-insensitive search
                    foreach (var c in sourceObj.GetComponents<Component>())
                    {
                        if (c != null && c.GetType().Name.Equals(componentType, StringComparison.OrdinalIgnoreCase))
                        {
                            sourceComp = c;
                            break;
                        }
                    }
                }

                if (sourceComp == null)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Component of type '{componentType}' not found on source GameObject '{sourceObj.name}'",
                        "component_not_found"
                    );
                }

                // Copy component to internal clipboard
                if (!UnityEditorInternal.ComponentUtility.CopyComponent(sourceComp))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Failed to copy component '{componentType}' from '{sourceObj.name}'",
                        "copy_failed"
                    );
                }

                // Paste on target
                Component targetComp = targetObj.GetComponent(sourceComp.GetType());
                bool alreadyExists = (targetComp != null);

                if (alreadyExists)
                {
                    if (UnityEditorInternal.ComponentUtility.PasteComponentValues(targetComp))
                    {
                        McpLogger.LogInfo($"[MCP Unity] Pasted component values of '{componentType}' from '{sourceObj.name}' to '{targetObj.name}'");
                        return new JObject
                        {
                            ["success"] = true,
                            ["message"] = $"Successfully pasted component values of '{componentType}' to '{targetObj.name}'",
                            ["action"] = "pasted_values"
                        };
                    }
                }
                else
                {
                    if (UnityEditorInternal.ComponentUtility.PasteComponentAsNew(targetObj))
                    {
                        McpLogger.LogInfo($"[MCP Unity] Added new component '{componentType}' from '{sourceObj.name}' to '{targetObj.name}'");
                        return new JObject
                        {
                            ["success"] = true,
                            ["message"] = $"Successfully added new component '{componentType}' to '{targetObj.name}'",
                            ["action"] = "added_new_component"
                        };
                    }
                }

                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Failed to paste component '{componentType}' on '{targetObj.name}'",
                    "paste_failed"
                );
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error copying component: {ex.Message}",
                    "copy_error"
                );
            }
        }

        /// <summary>
        /// Resolve one end of the copy via the shared resolver, so instance IDs use the same
        /// public-ID scheme every other tool hands out (synthetic under MCP_UNITY_ENTITY_ID_API,
        /// raw instance IDs otherwise) and paths reach inactive objects in every loaded scene.
        /// </summary>
        private GameObjectResolver.Result ResolveGameObject(JObject parameters, string role, string instanceIdKey, string pathKey)
        {
            int? instanceId = parameters[instanceIdKey]?.ToObject<int?>();
            string path = parameters[pathKey]?.ToObject<string>();

            if (!instanceId.HasValue && string.IsNullOrEmpty(path))
            {
                return new GameObjectResolver.Result
                {
                    Error = McpUnitySocketHandler.CreateErrorResponse(
                        $"{role} GameObject not specified. Provide a valid '{instanceIdKey}' or '{pathKey}'.",
                        "validation_error"
                    )
                };
            }

            return GameObjectResolver.Find(instanceId, path);
        }
    }
}
